File: research/features/l1.py
"""
L1 Feature Generator for Dollar Bar data.

Generates basic features from Adaptive Dynamic Dollar Bars:
- Log transforms for normalization
- Volume imbalance (buy/sell pressure)
- VWAP deviation
- Duration features (time-aware for irregular bars)
"""

import numpy as np
import pandas as pd
from typing import List, Optional


class L1FeatureGenerator:
    """
    Generate L1 (Level 1) features from Dollar Bar data.

    Expected input columns (from ETL Parquet):
        - start_time, end_time: int64 (ms)
        - open, high, low, close: float64
        - volume: float64 (BTC)
        - dollar_value: float64 (USD)
        - tick_count: int64
        - duration: float64 (seconds)
        - buy_dollar_vol, sell_dollar_vol: float64
        - net_imbalance: float64
    """

    def __init__(self, epsilon: float = 1e-10):
        """
        Args:
            epsilon: Small value to prevent division by zero
        """
        self.epsilon = epsilon

    def generate(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        Generate L1 features from Dollar Bar data.

        Args:
            df: DataFrame with Dollar Bar columns

        Returns:
            DataFrame with original columns + L1 features
        """
        features = df.copy()

        # 1. Log Transforms (normalize skewed distributions)
        # Note: log_dollar_value removed - constant in Dollar Bars (threshold-based)
        features['log_volume'] = np.log1p(features['volume'])
        features['log_tick_count'] = np.log1p(features['tick_count'])
        features['log_duration'] = np.log1p(features['duration'])

        # 2. VWAP Calculation and Deviation
        # VWAP = Total Dollar Value / Total Volume
        features['vwap'] = features['dollar_value'] / (features['volume'] + self.epsilon)
        features['vwap_deviation'] = (features['close'] - features['vwap']) / (features['close'] + self.epsilon)

        # 3. Volume Imbalance (normalized)
        # Already have net_imbalance = buy_dollar_vol - sell_dollar_vol
        # Normalize by total dollar value
        features['volume_imbalance'] = features['net_imbalance'] / (features['dollar_value'] + self.epsilon)

        # 4. Buy Ratio (proportion of buy volume)
        features['buy_ratio'] = features['buy_dollar_vol'] / (features['dollar_value'] + self.epsilon)

        # 5. Price Range Features
        features['bar_range'] = (features['high'] - features['low']) / (features['close'] + self.epsilon)
        features['bar_body'] = (features['close'] - features['open']) / (features['close'] + self.epsilon)

        # 6. Trade Intensity (dollars per second)
        features['trade_intensity'] = features['dollar_value'] / (features['duration'] + self.epsilon)
        features['log_trade_intensity'] = np.log1p(features['trade_intensity'])

        # 7. Tick Size (average trade size)
        features['avg_trade_size'] = features['dollar_value'] / (features['tick_count'] + self.epsilon)
        features['log_avg_trade_size'] = np.log1p(features['avg_trade_size'])

        return features

    def get_feature_names(self) -> List[str]:
        """Return list of generated feature names."""
        return [
            'log_volume',
            'log_tick_count',
            'log_duration',
            'vwap',
            'vwap_deviation',
            'volume_imbalance',
            'buy_ratio',
            'bar_range',
            'bar_body',
            'trade_intensity',
            'log_trade_intensity',
            'avg_trade_size',
            'log_avg_trade_size',
        ]

    def get_core_features(self) -> List[str]:
        """Return list of core features for model input."""
        return [
            'log_volume',
            'log_duration',
            'volume_imbalance',
            'vwap_deviation',
            'log_tick_count',
        ]

File: etl/internal/features/l1.go
package features

import (
	"math"
)

// L1Features contains the first-level (stateless) features.
// These are simple transformations that don't require historical state.
type L1Features struct {
	LogVolume         float64
	LogTickCount      float64
	LogDuration       float64
	LogTradeIntensity float64
	VWAPDeviation     float64
	VolumeImbalance   float64
	BarRange          float64
	BarBody           float64
	KyleLambda        float64 // Amihud illiquidity measure
}

// L1 computes stateless L1 features from bar data.
// All computations are O(1) with no memory allocation.
type L1 struct{}

// NewL1 creates a new L1 feature generator.
func NewL1() *L1 {
	return &L1{}
}

// Compute calculates all L1 features for a single bar.
// This is a pure function with no side effects.
func (l *L1) Compute(bar InputBar) L1Features {
	var f L1Features

	// === Log Transforms ===
	// Log1p(x) = log(1 + x), handles x=0 gracefully
	f.LogVolume = math.Log1p(bar.Volume)
	f.LogTickCount = math.Log1p(float64(bar.TickCount))
	f.LogDuration = math.Log1p(bar.Duration)

	// Trade Intensity = DollarValue / Duration
	// Measures how fast money is flowing through the market
	duration := bar.Duration
	if duration < Epsilon {
		duration = Epsilon
	}
	tradeIntensity := bar.DollarValue / duration
	f.LogTradeIntensity = math.Log1p(tradeIntensity)

	// === VWAP Deviation ===
	// VWAP = DollarValue / Volume (Volume-Weighted Average Price)
	// Deviation = (Close - VWAP) / Close
	// Positive = Close above VWAP (bullish), Negative = below (bearish)
	volume := bar.Volume
	if volume < Epsilon {
		volume = Epsilon
	}
	vwap := bar.DollarValue / volume
	f.VWAPDeviation = (bar.Close - vwap) / (bar.Close + Epsilon)

	// === Volume Imbalance ===
	// NetImbalance = BuyDollarVol - SellDollarVol
	// Imbalance = NetImbalance / TotalDollarValue
	// Range: [-1, 1], Positive = buy pressure, Negative = sell pressure
	dollarValue := bar.DollarValue
	if dollarValue < Epsilon {
		dollarValue = Epsilon
	}
	f.VolumeImbalance = bar.NetImbalance / dollarValue

	// === Bar Range ===
	// Range = (High - Low) / Close
	// Measures volatility within the bar, normalized by price
	close := bar.Close
	if close < Epsilon {
		close = Epsilon
	}
	f.BarRange = (bar.High - bar.Low) / close

	// === Bar Body ===
	// Body = (Close - Open) / Close
	// Positive = bullish (closed higher), Negative = bearish
	f.BarBody = (bar.Close - bar.Open) / close

	return f
}

// FeatureNames returns the names of all L1 features.
func (l *L1) FeatureNames() []string {
	return []string{
		"log_volume",
		"log_tick_count",
		"log_duration",
		"log_trade_intensity",
		"vwap_deviation",
		"volume_imbalance",
		"bar_range",
		"bar_body",
	}
}
